File: Scala/algorithmic_toolbox/src/main/scala-2.12/greedy/CollectingSignatures.scala
package greedy

import scala.annotation.tailrec

object CollectingSignatures {
  final case class Segment(left: Int, right: Int)

  private def convertToIntList(line: String): List[Int] = line.split(" ").map(_.toInt).toList

  private def getSegment(line: String): Segment = {
    val List(a, b): List[Int] = convertToIntList(line)
    Segment(a, b)
  }

  def calcMinimumNumberOfPointsCoveringSegments(segments: List[Segment]): List[Int] = {
    @tailrec
    def loop(remainingSegments: List[Segment], points: List[Int]): List[Int] = remainingSegments match {
      case Nil => points
      case Segment(_, b) :: rest => loop(rest.filter(_.left > b), b :: points)
    }
    loop(segments.sortBy(_.right), Nil)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val nrSegments :Int = reader.next().toInt
    val segments: List[Segment] = (for { _ <- 0 until nrSegments } yield getSegment(reader.next())).toList
    val covering: List[Int] = calcMinimumNumberOfPointsCoveringSegments(segments)
    println(covering.length)
    println(covering.mkString(" "))
  }
}

File: OCaml/algorithmic_toolbox/greedy/collecting_signatures.ml
type segment = {left : int; right : int}

let read_segments nr_segments =
    let pair_to_segment left right = {left; right} in
    let rec read_segment n acc = match n with
        | 0 -> acc
        | _ -> read_segment (n - 1) ((Scanf.sscanf (read_line ()) "%d %d" pair_to_segment) :: acc)
    in read_segment nr_segments []


let calc_minimum_number_of_points_covering_segments segments =
    let sorted_segments = List.sort (fun s1 s2 -> s1.right - s2.right) segments in
    let rec inner remaining_segments points = match remaining_segments with
        | [] -> points
        | {right} :: rest -> inner (List.filter (function {left} -> left > right) rest) (right :: points)
    in inner sorted_segments []


let () =
    let nr_segments = read_int() in
    let segments = read_segments nr_segments in
    let covering = calc_minimum_number_of_points_covering_segments segments in
    print_int (List.length covering);
    print_newline();
    List.iter (function elem -> print_int elem; print_string " ") covering;
    print_newline()
